### PR TITLE
James 1736 Introduce MessageParser which retrieve the list of attachments of a message

### PR DESCRIPTION
**Pull request by rouazana**
Fix remaining comments of #233

### CODE REVIEW

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        List<Attachment> attachments;
+        Body body = new DefaultMessageBuilder()
+                .parseMessage(fullContent)
+                .getBody();
+        if (body instanceof Multipart) {
+            attachments = listAttachments((Multipart)body);
+        } else {
+            attachments = ImmutableList.of();
+        }
+        body.dispose();
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: should be in a finally

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+public class MessageParser {
+
+    public List<Attachment> retrieveAttachments(InputStream fullContent) throws MimeException, IOException {
+        List<Attachment> attachments;
+        Body body = new DefaultMessageBuilder()
+                .parseMessage(fullContent)
+                .getBody();
+        if (body instanceof Multipart) {
+            attachments = listAttachments((Multipart)body);
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: why don't you return directly ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        body.dispose();
+        return attachments;
+    }
+
+    private List<Attachment> listAttachments(Multipart multipart) throws IOException {
+        ImmutableList.Builder<Attachment> attachments = ImmutableList.builder();
+        MessageWriter messageWriter = new DefaultMessageWriter();
+        for (Entity entity : multipart.getBodyParts()) {
+            if (isAttachment(entity)) {
+                attachments.add(createAttachment(messageWriter, entity.getBody()));
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: you don't need to dispose the body ?

Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: no, Multipart.dispose() is disposing each part.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 1 above)
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: ok

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 2 above)
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: now possible with the finally, else needed to dispose after

**Comment by mbaechler** [comment]
:+1:

**Comment by chibenwa** [comment]
:+1: (even if I maintain I would like to see at least one assert on the attachement content)

**Comment by aduprat** [comment]
:+1:

**Comment by rouazana** [comment]
@chibenwa I added your test case, please review again @all.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java [hunk 4]
```diff
@@ -51,6 +51,14 @@ public void getAttachmentsShouldRetrieveAttachmentsWhenOne() throws Exception {
...
+    public void getAttachmentsShouldReturnTheExpectedAttachment() throws Exception {
+        List<Attachment> attachments = testee.retrieveAttachments(ClassLoader.getSystemResourceAsStream("eml/oneAttachmentAndSomeInlined.eml"));
+
+        Attachment attachment = attachments.get(0);
+        assertThat(attachment.getStream()).hasContentEqualTo(ClassLoader.getSystemResourceAsStream("eml/exploits_of_a_mom.png"));
+    }
+
+    @Test
     public void getAttachmentsShouldRetrieveAttachmentsWhenTwo() throws Exception {
         List<Attachment> attachments = testee.retrieveAttachments(ClassLoader.getSystemResourceAsStream("eml/twoAttachments.eml"));
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java]: mmmh From XKCD licensing : 

```
This work is licensed under a Creative Commons Attribution-NonCommercial 2.5 License.

This means that you are free to copy and reuse any of my drawings (noncommercially) as long as you tell people where they're from.
```

Sorry :-(

Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java]: This was embedded in one of the emails. Do you think we should find other emails examples?

Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java]: Maybe, yes....

**Comment by rouazana** [comment]
I changed the offending file (replaced by a gimp icon, hoping it is free enough ;)), I'll open a ticket to review licence of other attachments.

**Comment by mbaechler** [comment]
:+1:

**Comment by aduprat** [comment]
:+1:

**Pull request merged**